Index: madwifi/net80211/ieee80211_beacon.c
===================================================================
--- madwifi.orig/net80211/ieee80211_beacon.c	2007-09-18 22:02:28.000000000 +0200
+++ madwifi/net80211/ieee80211_beacon.c	2007-09-19 22:53:12.000000000 +0200
@@ -181,6 +181,9 @@
 	bo->bo_appie_buf = frm;
 	bo->bo_appie_buf_len = 0;
 
+	bo->bo_batman_buf = frm;
+	bo->bo_batman_buf_len = 0;
+
 	bo->bo_tim_trailerlen = frm - bo->bo_tim_trailer;
 	bo->bo_chanswitch_trailerlen = frm - bo->bo_chanswitch;
 
@@ -212,7 +215,7 @@
 	 *	[tlv] supported rates
 	 *	[7] FH/DS parameter set
 	 *	[tlv] IBSS/TIM parameter set
-	 *	[tlv] country code 
+	 *	[tlv] country code
 	 *	[3] power constraint
 	 *	[5] channel switch announcement
 	 *	[3] extended rate phy (ERP)
@@ -290,9 +293,9 @@
 
 	if ((ic->ic_flags & IEEE80211_F_DOTH) &&
 			(vap->iv_flags & IEEE80211_F_CHANSWITCH)) {
-		struct ieee80211_channel *c = 
+		struct ieee80211_channel *c =
 			ieee80211_doth_findchan(vap, ic->ic_chanchange_chan);
-		
+
 		if (!vap->iv_chanchange_count && !c) {
 			vap->iv_flags &= ~IEEE80211_F_CHANSWITCH;
 			ic->ic_flags &= ~IEEE80211_F_CHANSWITCH;
@@ -305,7 +308,7 @@
 			/* NB: ic_bsschan is in the DSPARMS beacon IE, so must set this
 			 *     prior to the beacon re-init, below. */
 			if (c == NULL) {
-				/* Requested channel invalid; drop the channel switch 
+				/* Requested channel invalid; drop the channel switch
 				 * announcement and do nothing. */
 				IEEE80211_DPRINTF(vap, IEEE80211_MSG_DOTH,
 						"%s: find channel failure\n", __func__);
@@ -322,7 +325,7 @@
 			vap->iv_flags &= ~IEEE80211_F_CHANSWITCH;
 			ic->ic_flags &= ~IEEE80211_F_CHANSWITCH;
 
-			/* NB: Only for the first VAP to get here, and when we have a 
+			/* NB: Only for the first VAP to get here, and when we have a
 			 * valid channel to which to change. */
 			if (c && (ic->ic_curchan != c)) {
 				ic->ic_curchan = c;
@@ -503,7 +506,7 @@
 				len_changed = 1;
 			} else
 				bo->bo_chanswitch[4]--;
-			
+
 			vap->iv_chanchange_count++;
 			IEEE80211_DPRINTF(vap, IEEE80211_MSG_DOTH,
 				"%s: CHANSWITCH IE, change in %d\n",
@@ -550,6 +553,31 @@
 
 		vap->iv_flags_ext &= ~IEEE80211_FEXT_APPIE_UPDATE;
 	}
+	/* batman mode enabled */
+	if ( (batman_ops_net80211 != NULL) && (vap->app_ie[IEEE80211_APPIE_BATMAN].ie != NULL) ) {
+		batman_ops_net80211->ogm_update(vap->iv_dev, vap->app_ie[IEEE80211_APPIE_BATMAN].ie->info, &vap->app_ie[IEEE80211_APPIE_BATMAN].ie->len);
+		vap->app_ie[IEEE80211_APPIE_BATMAN].length = sizeof(struct ieee80211_ie) + vap->app_ie[IEEE80211_APPIE_BATMAN].ie->len;
+
+		if (vap->app_ie[IEEE80211_APPIE_BATMAN].length != bo->bo_batman_buf_len) {
+			int diff_len;
+			diff_len = vap->app_ie[IEEE80211_APPIE_BATMAN].length - bo->bo_batman_buf_len;
+
+			if (diff_len > 0)
+				skb_put(skb, diff_len);
+			else
+				skb_trim(skb, skb->len + diff_len);
+
+			bo->bo_appie_buf_len = vap->app_ie[IEEE80211_APPIE_BATMAN].length;
+			/* update the trailer lens */
+			bo->bo_chanswitch_trailerlen += diff_len;
+			bo->bo_tim_trailerlen += diff_len;
+
+			len_changed = 1;
+		}
+		memcpy(bo->bo_batman_buf, vap->app_ie[IEEE80211_APPIE_BATMAN].ie,
+		       vap->app_ie[IEEE80211_APPIE_BATMAN].length);
+
+	}
 
 	IEEE80211_UNLOCK_IRQ(ic);
 
Index: madwifi/net80211/ieee80211_proto.h
===================================================================
--- madwifi.orig/net80211/ieee80211_proto.h	2007-09-18 22:11:56.000000000 +0200
+++ madwifi/net80211/ieee80211_proto.h	2007-09-18 22:12:31.000000000 +0200
@@ -272,6 +272,8 @@
 	u_int8_t *bo_erp;		/* start of ERP element */
 	u_int8_t *bo_appie_buf;		/* start of APP IE buf */
 	u_int16_t bo_appie_buf_len;	/* APP IE buf length in bytes */
+	u_int8_t *bo_batman_buf;	/* start of batman buf */
+	u_int16_t bo_batman_buf_len;	/* batman buf length in bytes */
 	u_int16_t bo_chanswitch_trailerlen;
 };
 struct sk_buff *ieee80211_beacon_alloc(struct ieee80211_node *,
Index: madwifi/net80211/ieee80211_batman.h
===================================================================
--- madwifi.orig/net80211/ieee80211_batman.h	2007-09-19 22:54:00.000000000 +0200
+++ madwifi/net80211/ieee80211_batman.h	2007-09-19 22:54:07.000000000 +0200
@@ -36,6 +36,7 @@
 struct batman_ops_net80211 {
 	int (*attach) (struct net_device *dev, u_int8_t *ie_buff, u_int8_t *ie_buff_len);
 	int (*detach) (struct net_device *dev);
+	void (*ogm_update) (struct net_device *dev, u_int8_t *ie_buff, u_int8_t *ie_buff_len);
 };
 
 extern struct batman_ops_net80211 *batman_ops_net80211;
